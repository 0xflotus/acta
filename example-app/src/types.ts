export interface ITodo {
  label: string;
  date: number;
}
