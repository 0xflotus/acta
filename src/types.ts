import React from 'react';

export type TActaValue = string | number | object | boolean | null | undefined;

export interface IState {
  value: TActaValue;
  defaultValue: TActaValue;
  subscribtions: {
    [contextID: string]: {
      callback: (value: TActaValue) => void;
      context?: IComponentWithID;
    };
  };
}

export interface IEvent {
  [contextID: string]: {
    callback: (value: TActaValue) => void;
    context: IComponentWithID;
  };
}

export interface IComponentWithID extends React.Component {
  actaID?: string;
}

export interface IActa {
  initialized: boolean;

  states: {
    [actaStateKey: string]: IState;
  };
  events: {
    [actaEventKey: string]: IEvent;
  };
  actaIDs: Array<string>;

  init: (userProvidedWindow?: Window) => void;

  ensureActaID: (context: IComponentWithID) => boolean | string;

  useActaState: (actaStateKey: string, defaultValue?: TActaValue) => TActaValue;

  subscribeState: (
    actaStateKey: string,
    callbackOrLocalStateKey: string | ((valueToReturn: any) => void),
    context: IComponentWithID,
    defaultValue?: TActaValue
  ) => void;

  unsubscribeState: (stateKey: string, context: IComponentWithID) => void;

  setState: (
    states: {
      [stateKey: string]: TActaValue;
    },
    persistenceType?: 'sessionStorage' | 'localStorage'
  ) => void;

  getState: (stateKey: string) => TActaValue;

  hasState: (stateKey: string) => boolean;

  deleteState: (
    stateKey: string,
    persistenceType?: 'localStorage' | 'sessionStorage'
  ) => void;

  subscribeEvent: (
    eventKey: string,
    callback: (valueToReturn: any) => void,
    context: IComponentWithID
  ) => void | boolean;

  unsubscribeEvent: (eventKey: string, context: IComponentWithID) => void;

  dispatchEvent: (
    eventKey: string,
    data?: TActaValue,
    isShared?: boolean
  ) => void;
}
