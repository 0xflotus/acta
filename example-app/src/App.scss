#main {
  padding-bottom: 100px;

  > header {
    > h1 {
      font-size: 32px;
      font-weight: bold;
      line-height: 1;
      width: 300px;
      padding: 20px;
    }
  }

  > .todoLists {
    border-bottom: 1px solid #000;
    display: flex;
    flex-direction: column;

    @media (min-width: 800px) {
      flex-direction: row;
    }

    > div {
      border-top: 1px solid #000;
      flex-grow: 0;
      flex-shrink: 0;

      @media (min-width: 800px) {
        width: 50%;

        &:first-of-type {
          border-right: 1px solid #000;
        }
      }

      > h2 {
        border-bottom: 1px solid #000;
        font-size: 24px;
        font-weight: bold;
        padding: 20px;
      }

      > p {
        color: orange;
        padding: 20px;
      }

      > div {
        display: flex;
        border-bottom: 1px solid #000;
        align-items: stretch;
        justify-content: space-between;

        &:last-of-type {
          border-bottom: none;
        }

        > div {
          padding: 10px;

          > span {
            display: block;

            &:first-of-type {
              font-size: 14px;
            }

            &:last-of-type {
              color: #666;
              font-size: 9px;
            }
          }
        }

        > button {
          border: none;
          padding: 10px 20px;
          font-weight: bold;
          text-transform: uppercase;
          font-size: 10px;
          background: #eee;
          letter-spacing: 1px;
          transition: filter 200ms ease-in-out;

          &:hover {
            filter: brightness(0.75);
          }
        }
      }
    }
  }

  > form {
    padding: 20px;
    border-bottom: 1px solid #000;
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    background-color: #eee;

    > h3 {
      width: 100%;
      font-weight: bold;
      margin-bottom: 5px;
    }

    > input[type='text'] {
      display: block;
      width: 100%;
      border: 1px solid #000;
      padding: 10px;
      font-size: 14px;
      margin-bottom: 10px;
    }

    > p {
      font-size: 12px;
      display: flex;
      align-items: center;
      margin-bottom: 10px;

      > input[type='checkbox'] {
        display: block;
        margin-left: 5px;
        appearance: none;
        width: 15px;
        height: 15px;
        background-color: #fff;
        border: 1px solid #000;
        position: relative;
        cursor: pointer;

        &:checked {
          &:after {
            content: '';
            display: block;
            position: absolute;
            width: 11px;
            height: 11px;
            top: 1px;
            left: 1px;
            background-color: #000;
          }
        }
      }
    }

    > input[type='submit'] {
      border: none;
      background: #000;
      padding: 10px 20px;
      font-weight: bold;
      text-transform: uppercase;
      font-size: 10px;
      letter-spacing: 1px;
      cursor: pointer;
      color: #fff;
    }
  }

  > .messagesContainer {
    position: fixed;
    bottom: 0;
    z-index: 9;

    &.functional {
      left: 0;
    }

    &.class {
      right: 0;
    }

    > div {
      margin: 10px;
      padding: 10px;
      background-color: #eee;
      border: 1px solid #000;
      border-left: 5px solid #000;
    }
  }
}
