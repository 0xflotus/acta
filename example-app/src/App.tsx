import React from 'react';

import { FunctionalTodosList } from './components/FunctionalTodosList';
import { ClassTodoList } from './components/ClassTodoList';
import { AddTodoForm } from './components/AddTodoForm';

export const App = () => {
  return (
    <div className='App'>
      <FunctionalTodosList />
      <ClassTodoList />

      <AddTodoForm />
    </div>
  );
};
