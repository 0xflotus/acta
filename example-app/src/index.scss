* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  font: inherit;
  vertical-align: baseline;

  &:focus {
    outline: none;
  }
}

body {
  user-select: none;
  font-size: 16px;
  line-height: 1.5;
  text-rendering: optimizeLegibility;
  word-wrap: break-word;
  min-height: 100vh;
  overflow: hidden;
  font-family: Helvetica, 'Helvetica Neue', Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  > #root {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    overflow: auto;
  }
}

a,
button {
  cursor: pointer;
}

b {
  font-weight: 700;
}

i {
  font-style: italic;
}

::-moz-selection {
  background: #333;
  color: #fff;
}

::selection {
  background: #333;
  color: #fff;
}
