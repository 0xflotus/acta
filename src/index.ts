import { IActa } from './types';
import { isObject } from './isObject';

const actaStoragePrefix = '__acta__';
const actaStoragePrefixLength = actaStoragePrefix.length;
const actaEventPrefix = '__actaEvent__';
const actaEventPrefixLength = actaEventPrefix.length;
const isInDOM = typeof window !== 'undefined';

const Acta: IActa = {
  /**
   * Boolean to know if acta has been initialized before
   */
  initialized: false,

  /**
   * states will hold values, subscribers list and history indexed by keys
   * events will hold subscribers list indexed by events keys
   * actions will hold actions function indexed by actions keys
   * actaIDs will be an array of unique ids
   */
  actaIDs: [],
  events: {},
  states: {},

  /**
   * The init function will search for any previousely acta states in the local
   * and session storage and will start listening on the changes in the storages
   * to update itself if the storage is updated from another tab
   */
  init(): void {
    /**
     * Acta is now initialized
     */
    this.initialized = true;

    /**
     * Attach to the window for global public access
     */
    (window as any).Acta = Acta;

    /**
     * Listen to the storage to synchronise Acta between tabs
     */
    window.addEventListener(
      'storage',
      event => {
        if (
          event.key &&
          event.key.slice(0, actaStoragePrefixLength) === actaStoragePrefix
        ) {
          if (
            event.newValue !== null &&
            event.newValue !== '' &&
            event.newValue !== 'null'
          ) {
            this.setState({
              [event.key.slice(actaStoragePrefixLength)]: JSON.parse(
                event.newValue,
              ),
            });
          } else {
            this.setState({
              [event.key.slice(actaStoragePrefixLength)]: null,
            });
          }
        } else if (
          event.key &&
          event.key.slice(0, actaEventPrefixLength) === actaEventPrefix
        ) {
          this.dispatchEvent(
            event.key.slice(actaEventPrefixLength),
            event.newValue ? JSON.parse(event.newValue) : event.newValue,
            false,
          );
        }
      },
      false,
    );

    /**
     * Load the data from local and session storage
     */
    const storage = {
      ...localStorage,
      ...sessionStorage,
    };
    const storageKeys = Object.keys(storage);
    for (const storageKey of storageKeys) {
      if (storageKey.slice(0, actaStoragePrefixLength) === actaStoragePrefix) {
        const value = JSON.parse(storage[storageKey]);
        if (value !== undefined && value !== null) {
          this.setState({
            [storageKey.slice(actaStoragePrefixLength)]: value,
          });
        }
      }
    }
  },

  /**
   * subscribeState is called from a react component with a callback
   * when that state will be set, its value will be send
   * to the component by the callback
   *
   * that subsribtion can be destroyed manually and will be
   * destroyed automatically when the
   *
   * @param {String} stateKey - The key to name the state
   * @param {Function} callback - Reference to the callback that will
   * be called when the state change
   * @param {Object} context - Reference to the react component from
   * wich the subscribtion is made => that will be needed to unsubscribe
   * when the compnent will unmount
   * @param {TActaValue} defaultValue - Optionnal, set a default value for
   * the state if there is none
   */
  subscribeState(stateKey, callback, context, defaultValue) {
    /* Ensure the arguments */
    if (
      typeof stateKey !== 'string' ||
      typeof callback !== 'function' ||
      !isObject(context)
    ) {
      throw new Error(
        `Acta.subscribeState params =>
[0]: string,
[1]: function,
[2]: mounted react component`,
      );
    }

    /* If the component does not have an acta id, get him one */
    this.ensureActaID(context);

    /* If this state does not already exists, creates it */
    this.states[stateKey] = this.states[stateKey] || {
      value: defaultValue || undefined,
      defaultValue: defaultValue || undefined,
      subscribtions: {},
    };

    /**
     * If a subscribtion for this context on this state
     * already exists, stop here
     */
    if (this.states[stateKey].subscribtions[context.actaID as string])
      return false;

    /**
     * Extend the componentWillUnmount hook on the context
     * with a state unsubscribtion, so when
     * the component unmounts, all its states subscribtions
     * will be destroyed
     */
    if (context.componentWillUnmount) {
      const oldComponentWillUnmount = context.componentWillUnmount;
      context.componentWillUnmount = () => {
        this.unsubscribeState(stateKey, context);
        oldComponentWillUnmount.bind(context)();
      };
    } else {
      context.componentWillUnmount = () =>
        this.unsubscribeState(stateKey, context);
    }

    /**
     * Add the callback and the context to the subscribtion list
     * of the state
     */
    this.states[stateKey].subscribtions[context.actaID as string] = {
      callback,
      context,
    };

    /* Dispatch the initial or current value to the subscriber
		if initialize is not set to false and if there is a valid
		non circular state to dispatch */
    try {
      const valueToReturn = JSON.stringify(this.states[stateKey].value || null);
      callback(JSON.parse(valueToReturn));
      return JSON.parse(valueToReturn);
    } catch (err) {
      console.error(`Error in subscribeState for "${stateKey}":`, err);
      return false;
    }
  },

  /**
   * unsubscribeState will simply remove the target context form
   * the subscribtions of the target state
   *
   * @param {String} stateKey - The key to name the state
   * @param {Object} context - Reference to the target react component
   */
  unsubscribeState(stateKey, context) {
    /* Ensure the arguments */
    if (
      typeof stateKey !== 'string' ||
      !isObject(context) ||
      !this.states[stateKey]
    ) {
      throw new Error(
        `Acta.unsubscribeState params =>
[0]: string,
[2]: mounted react component`,
      );
    }

    /* Delete the subscribtion */
    delete this.states[stateKey].subscribtions[context.actaID as string];
  },

  /**
   * set state will save the value in the the tagret state
   * and will dispatch that update to all the subscriber
   * by he provided callback
   *
   * @param {Object} states - mandatory - an object where the keys are states to set and
   * values the target values
   * @param {String} persistenceType - optionnal - can be "sessionStorage" or "localStorage"
   * if set, the state will be saved into the corresponding storage
   */
  setState(states, persistenceType) {
    /* Ensure the arguments */
    if (!isObject(states) || Object.keys(states).length === 0) {
      throw new Error('Acta.setState params => [0]: object with 1+ key');
    }

    /**
     * Loop over the states
     */
    for (const state of Object.entries(states)) {
      const stateKey = state[0];
      const value = state[1];

      /* If this state does not already exists, creates it */
      if (!this.states[stateKey]) {
        this.states[stateKey] = this.states[stateKey] || {
          value: value || undefined,
          defaultValue: undefined,
          subscribtions: {},
        };
      }

      /* Save the value */
      this.states[stateKey].value = value;

      /* If persistence is configured and we have a window, store the value */
      if (isInDOM && persistenceType) {
        if (persistenceType === 'localStorage') {
          localStorage.setItem(
            `${actaStoragePrefix}${stateKey}`,
            JSON.stringify(value),
          );
        } else if (persistenceType === 'sessionStorage') {
          sessionStorage.setItem(
            `${actaStoragePrefix}${stateKey}`,
            JSON.stringify(value),
          );
        } else {
          throw new Error(
            'Acta.setState params => [1]: "sessionStorage" | "localStorage".',
          );
        }
      }

      /**
       * Try to dispatch to all subscribers & kill the
       * subscribtion if the subscriber has been destroyed
       */
      Object.keys(this.states[stateKey].subscribtions || {}).forEach(actaID => {
        try {
          this.states[stateKey].subscribtions[actaID].callback(value);
        } catch (err) {
          if (
            !this.states[stateKey].subscribtions[actaID]?.context ||
            !this.states[stateKey].subscribtions[actaID].callback
          ) {
            delete this.states[stateKey].subscribtions[actaID];
          }
        }
      });
    }
  },

  /**
   * Delete a state
   *
   * @param {String} state the state to target
   */
  deleteState(stateKey, persistenceType) {
    /* Ensure the arguments */
    if (typeof stateKey !== 'string' || stateKey === '') {
      throw new Error('Acta.deleteState params => [0]: string');
    }

    delete this.states[stateKey];

    /**
     * If the persistance type is set and we have a window, remove the
     * value from the storage
     */
    if (isInDOM && persistenceType) {
      if (persistenceType === 'sessionStorage') {
        sessionStorage.removeItem(`${actaStoragePrefix}${stateKey}`);
      } else if (persistenceType === 'localStorage') {
        localStorage.removeItem(`${actaStoragePrefix}${stateKey}`);
      } else {
        throw new Error(
          'Acta.deleteState params => [1]: "sessionStorage" | "localStorage".',
        );
      }
    }
  },

  /**
   * return the state form its name
   *
   * @param {String} stateKey - the key to identify the target state
   * @return {*} can be anything
   */
  getState(stateKey) {
    /* Check the parameter */
    if (typeof stateKey !== 'string' || !this.hasState(stateKey)) {
      throw new Error('Acta.deleteState params => [0]: string');
    }

    return (
      this.states[stateKey].value ||
      this.states[stateKey].defaultValue ||
      undefined
    );
  },

  /**
   * check the existence of a state
   *
   * @param {String} stateKey - the key to identify the target state
   */
  hasState(stateKey) {
    /**
     * Check param
     */
    if (typeof stateKey !== 'string') {
      throw new Error('Acta.hasState params => [0]: string');
    }
    return this.states.hasOwnProperty(stateKey);
  },

  /**
   * subscribeEvent is called from a react component with a callback
   * when the corresponding event is dispatched, the callback is called
   * passing any argument set in the dispatcher
   *
   * that subsribtion can be destroyed manually and will be
   * destroyed automatically when the component unmount
   *
   * @param {String} eventKey - The key to name the event
   * @param {Function} callback - Reference to the callback that will
   * be called when the state change
   * @param {React.Component} context - Reference to the react component from
   * wich the subscribtion is made => that will be needed to unsubscribe
   * when the compnent woll unmount
   */
  subscribeEvent(eventKey, callback, context): void {
    /* Ensure the arguments */
    if (
      typeof eventKey !== 'string' ||
      typeof callback !== 'function' ||
      !isObject(context)
    ) {
      throw new Error(
        `Acta.subscribeEvent params =>
[0]: string,
[1]: function,
[2]: mounted react component`,
      );
    }

    /* If this event does not already exists, creates it */
    this.events[eventKey] = this.events[eventKey] || {};

    /* If the component does not have an acta id, get him one */
    this.ensureActaID(context);

    /* If this context already listen to that event already exists, stop here */
    if (this.events[eventKey][context.actaID as string]) return;

    /**
     * Extend the componentWillUnmount hook on the context
     * with a event unsubscribtion, so when
     * the component unmounts, all its events subscribtions
     * will be destroyed
     */
    if (context.componentWillUnmount) {
      const oldComponentWillUnmount = context.componentWillUnmount;
      context.componentWillUnmount = () => {
        this.unsubscribeEvent(eventKey, context);
        oldComponentWillUnmount.bind(context)();
      };
    } else {
      context.componentWillUnmount = () => {
        this.unsubscribeEvent(eventKey, context);
      };
    }

    /**
     * Add the callback and the context to the event listener list
     * of the event
     */
    this.events[eventKey][context.actaID as string] = {
      callback,
      context,
    };
  },

  /**
   * unsubscribeEvent will remove the target context form
   * the subscribtions of the target eve,t
   *
   * @param {String} eventKey - The key to name the eventKey
   * @param {Object} context - Reference to the target react component
   */
  unsubscribeEvent(eventKey, context) {
    /* Ensure the arguments */
    if (
      typeof eventKey !== 'string' ||
      !isObject(context) ||
      !this.events[eventKey]
    ) {
      throw new Error(
        `Acta.subscribeEvent params =>
[0]: string,
[2]: mounted react component`,
      );
    }

    /* Delete the subscribtion */
    delete this.events[eventKey][context.actaID as string];
  },

  /**
   * Dispatch an event : check if the event exists and has subscribers
   * if it does, call the callback of each subscriber
   *
   * @param {String} eventKey - the key to target the event
   * @param {TActaValue} data - the data passed with the event
   * @param {Boolean} isShared - if the event should be shared accross tabs and windows
   */
  dispatchEvent(eventKey, data, isShared) {
    /* Ensure the arguments */
    if (typeof eventKey !== 'string' || !this.events[eventKey]) {
      throw new Error(
        'Acta.dispatchEvent params => [0]: string & must exist in Acta.events',
      );
    }

    /**
     * If this is a shared message, send it to the local storage;
     * it will come back instantly
     */
    if (isShared && isInDOM) {
      localStorage.setItem(
        `${actaEventPrefix}${eventKey}`,
        JSON.stringify(data),
      );
    }

    /**
     * Call each subscriber callback
     */
    Object.keys(this.events[eventKey]).forEach(actaID => {
      try {
        this.events[eventKey][actaID].callback(data || null);
      } catch (err) {
        if (
          !this.events[eventKey][actaID].context ||
          !this.events[eventKey][actaID].callback
        ) {
          delete this.events[eventKey][actaID];
        }
      }
    });
  },

  /**
   * ensureActaID is a small utility function to
   * inject a unique id to all subscribers contexts
   */
  ensureActaID(context) {
    /* Stops there if there is already an ID */
    if (!context || context.actaID) return false;

    /* Insert a new ID in the list and inject it int the context */
    const newId = `_${this.actaIDs.length}`;
    this.actaIDs.unshift(newId);
    context.actaID = newId;
    return newId;
  },
};

/**
 * If Acta has not been initialized, init Acta
 */
if (!Acta.initialized && isInDOM) Acta.init();

export default Acta;
