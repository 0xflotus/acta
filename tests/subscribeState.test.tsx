// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React from 'react';
import renderer from 'react-test-renderer';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import App, {
  addElementsInActa,
  ACTA_STATE_ELEMENTS_LIST_WITH_CALLBACK,
} from './testApp';
import Acta from '../src';
import { IComponentWithID } from '../src/types';

const paramsErrorMessage = `Acta.subscribeState params =>
[0]: string,
[1]: function or string,
[2]: mounted react component`;

describe('Acta subscribeState.test method', () => {
  /**
   * Feature
   */
  test('When a state change, the callback is triggered in the component.', () => {
    const app = renderer.create(<App />);

    // Render the app with an empty elements array
    let tree = app.toJSON();
    // The ul child should have no children
    let children = tree?.children?.find(
      // @ts-ignore
      (child) => child.props.id === 'listContainer'
      // @ts-ignore
    )?.children;
    expect(children).toBe(null);
    // Snapshot the resulting markup
    expect(tree).toMatchSnapshot();

    // Set the value in Acta state
    // It should trigger the callback
    addElementsInActa();

    // Render the app with an elements array containing strings now
    tree = app.toJSON();
    // The ul child should have children now
    children = tree?.children?.find(
      // @ts-ignore
      (child) => child.props.id === 'listContainer'
      // @ts-ignore
    )?.children;
    expect(Array.isArray(children)).toBe(true);
    expect(children.length > 0).toBe(true);
    // Snapshot the resulting markup
    expect(tree).toMatchSnapshot();
  });

  test('When a component has already subscribed to a state, a new subscritpion return false', () => {
    const actaSubs =
      Acta.states[ACTA_STATE_ELEMENTS_LIST_WITH_CALLBACK].subscribtions;
    const alreadySubscribedContext = actaSubs[Object.keys(actaSubs)[0]]
      .context as IComponentWithID;

    expect(
      Acta.subscribeState(
        ACTA_STATE_ELEMENTS_LIST_WITH_CALLBACK,
        () => true,
        alreadySubscribedContext
      )
    ).toBeUndefined();
  });

  test('When a component unmounts, it should not be in the acta subs anymore', () => {
    const actaSubs =
      Acta.states[ACTA_STATE_ELEMENTS_LIST_WITH_CALLBACK].subscribtions;
    const alreadySubscribedContext = actaSubs[Object.keys(actaSubs)[0]]
      .context as IComponentWithID;

    // We should have one sub
    expect(Object.keys(actaSubs).length).toBe(1);

    if (
      alreadySubscribedContext &&
      alreadySubscribedContext.componentWillUnmount
    ) {
      alreadySubscribedContext?.componentWillUnmount();
    }

    // We should not has subs anymore
    expect(Object.keys(actaSubs).length).toBe(0);
  });

  /**
   * Errors management
   */
  test('When the stateKey param is not a valid string, should throw an error', () => {
    expect(() => {
      // @ts-ignore
      Acta.subscribeState(null, () => true, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('', () => true, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState([], () => true, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState({}, () => true, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      Acta.subscribeState(
        // @ts-ignore
        () => true,
        // @ts-ignore
        () => true,
        // @ts-ignore
        {} as IComponentWithID
      );
    }).toThrowError(paramsErrorMessage);
  });

  test('When the callback param is not a valid function, should throw an error', () => {
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', null, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', undefined, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', [], {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', {}, {} as IComponentWithID);
    }).toThrowError(paramsErrorMessage);
  });

  test('When the context is not param is not an object, should throw an error', () => {
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', () => true);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', () => true, '');
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', () => true, []);
    }).toThrowError(paramsErrorMessage);
    expect(() => {
      // @ts-ignore
      Acta.subscribeState('testKey', () => true, null);
    }).toThrowError(paramsErrorMessage);
  });

  test('When the initial value is an object containing a circular object, should return false', () => {
    const circularObject: { item?: object } = {};
    circularObject.item = circularObject;

    expect(
      Acta.subscribeState(
        'testKey',
        () => true,
        {} as IComponentWithID,
        circularObject
      )
    ).toBe(undefined);
  });
});
