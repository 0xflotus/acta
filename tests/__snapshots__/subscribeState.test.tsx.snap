// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Acta subscribeState.test method When a state change, the callback is triggered in the component. 1`] = `
<div>
  <h1>
    Elements
  </h1>
  <ul
    id="listContainer"
  />
</div>
`;

exports[`Acta subscribeState.test method When a state change, the callback is triggered in the component. 2`] = `
<div>
  <h1>
    Elements
  </h1>
  <ul
    id="listContainer"
  >
    <li>
      a
    </li>
    <li>
      b
    </li>
  </ul>
</div>
`;
